mayastor/src/rebuild.rs

```diff
@@ -1,284 +1,224 @@
-#![allow(dead_code)]
-
-use std::{os::raw::c_void, time::SystemTime};
-
-use futures::channel::oneshot;
-use nix::errno::Errno;
-use snafu::Snafu;
-
-use spdk_sys::spdk_bdev_io;
-
 use crate::{
-    bdev::nexus::nexus_bdev::{nexus_lookup, NexusState},
-    core::{Descriptor, DmaBuf, DmaError},
+    bdev::nexus::nexus_bdev::nexus_lookup,
+    core::{Bdev, BdevHandle, CoreError, DmaBuf, DmaError, Reactors},
 };
+use snafu::{ResultExt, Snafu};
 
 #[derive(Debug, Snafu)]
-pub enum Error {
-    #[snafu(display(
-        "Rebuild source {} is larger than the target {}",
-        src,
-        tgt
-    ))]
-    SourceBigger { src: String, tgt: String },
-    #[snafu(display("Cannot suspend an already completed rebuild task"))]
-    SuspendCompleted {},
-    #[snafu(display(
-        "Cannot resume a rebuild task which has not been suspended"
-    ))]
-    ResumeNotSuspended {},
-    #[snafu(display("Failed to dispatch rebuild IO"))]
-    DispatchIo { source: Errno },
-    #[snafu(display("Failed to allocate buffer for rebuild IO"))]
-    BufferAlloc { source: DmaError },
+#[snafu(visibility = "pub(crate)")]
+pub enum RebuildError {
+    #[snafu(display("Failed to allocate buffer for the rebuild copy"))]
+    NoCopyBuffer { source: DmaError },
+    #[snafu(display("Failed to validate rebuild task creation parameters"))]
+    InvalidParameters {},
+    #[snafu(display("Failed to get a handle for bdev {}", bdev))]
+    NoBdevHandle { source: CoreError, bdev: String },
+    #[snafu(display("IO failed for bdev {}", bdev))]
+    IoError { source: CoreError, bdev: String },
 }
 
-#[derive(Debug, Copy, Clone, PartialOrd, PartialEq)]
+#[derive(Debug, PartialEq)]
 pub enum RebuildState {
-    /// Initialized properly, but not running
-    Initialized,
-    /// Running
+    Pending,
     Running,
-    /// Completed
-    Completed,
-    /// the rebuild task has failed
     Failed,
-    /// a request has been made to cancel the task
-    Cancelled,
-    /// Suspended
-    Suspended,
+    Completed,
 }
 
-/// struct that holds the state of a copy task. This struct
-/// is used during rebuild.
 #[derive(Debug)]
 pub struct RebuildTask {
+    nexus_name: String,
+    source: String,
```

Comment:
If we instead of storing Strings and BdevHandles, simply store the Descriptor, we can get the name and handle *from* the descriptor.

iirc. Descriptor as a method .get_bdev() which in turn has .get_name(). The name though is not really important to us outside of logging or things like that. 

I think we discussed this yesterday, so we can leave it for now, but please do file an issue for this.